core/env/shared/update.rb

```diff
@@ -19,19 +19,42 @@
     ENV.send(@method, {"foo" => "0", "bar" => "1"}).should equal(ENV)
   end
 
-  it "yields key, the old value and the new value when replacing entries" do
+  it "yields key, the old value and the new value when replacing an entry" do
     ENV.send @method, {"foo" => "0", "bar" => "3"}
     a = []
     ENV.send @method, {"foo" => "1", "bar" => "4"} do |key, old, new|
       a << [key, old, new]
+      new
+    end
+    a[0].should == ["foo", "0", "1"]
+    a[1].should == ["bar", "3", "4"]
+  end
+
+  it "yields key, the old value and the new value when replacing an entry" do
+    ENV.send @method, {"foo" => "0", "bar" => "3"}
+    ENV.send @method, {"foo" => "1", "bar" => "4"} do |key, old, new|
       (new.to_i + 1).to_s
     end
     ENV["foo"].should == "2"
     ENV["bar"].should == "5"
-    a[0].should == ["foo", "0", "1"]
-    a[1].should == ["bar", "3", "4"]
   end
 
+  # BUG: This fails because the block is evaluated.
+  # it "does not evaluate the block for a new name" do
+  #   ENV.delete("bar")
+  #   ENV.send @method, {"foo" => "0"}
+  #   ENV.send(@method, "bar" => "1") { |key, old, new| fail "Should not get here" }
+  # end
+
+  # BUG: This fails because the block's return value becomes the new name.
+  # it "does not use the block's return value as the value for a new name" do
+  #   ENV.delete("bar")
+  #   ENV.send @method, {"foo" => "0"}
+  #   ENV.send(@method, "bar" => "1") { |key, old, new| "Should not use this value" }
+  #   ENV["foo"].should == "0"
+  #   ENV["bar"].should == "1"
+  # end
```

Comment:
If we know this will be the behavior after these bugs are fixed, then it would be good to use `ruby_version_is "2.7" do`.

Reply:
In any case we should link to the relevant MRI bug for each of these specs.